### PR TITLE
Adding Schedule Messaging functionality

### PR DESCRIPTION
**Pull request by MattDavisRV**
Slack recently introduced schedule messages to their platform.  This PR adds support for this functionality.

More information can be found here: https://api.slack.com/messaging/scheduling

This PR includes the ability to schedule messages, list messages that are currently scheduled by channel, and delete scheduled messages.  

Note: During testing, Slack will not let you delete a message if it's being sent "soon".  I'm not exactly sure what the timing is, but it seems within 30 seconds of the scheduled time, the response will say your ScheduledMessageID is invalid.

I tried to follow the existing PostMessage style when adding this feature.

Any and all feedback welcome.

### CODE REVIEW

File: chat.go [hunk 1]
```diff
@@ -435,6 +545,50 @@ func MsgOptionIconEmoji(iconEmoji string) MsgOption {
 	}
 }
 
+// MsgOptionCursor sets the cursor value when searching foor scheduled messages
```
Comment (james-lawrence) [inline chat.go]: s/foor/for/

File: chat.go [hunk 2]
```diff
@@ -173,6 +244,33 @@ func (api *Client) SendMessageContext(ctx context.Context, channelID string, opt
...
+}
+
+// DeleteScheduledMessageContext removes a scheduled message with custom context.
+// See DeleteScheduledMessage for more information
+func (api *Client) DeleteScheduledMessageContext(ctx context.Context, channelID, scheduledMessageID string) (string, string, error) {
+	respChannel, respTimestamp, _, err := api.SendMessageContext(
+		ctx,
+		channelID,
+		MsgOptionDeleteSchedule(scheduledMessageID),
+		MsgOptionCompose(),
```
Comment (james-lawrence) [inline chat.go]: eh? this option doesn't do anything?

**Review by james-lawrence — COMMENTED** [review]
larger PR than I expected. will have to wait a bit for a full review.

File: chat.go (hunk 1 above)
Comment (MattDavisRV) [reply chat.go]: My keyboard has seen better days.. this has been fixed.

File: chat.go (hunk 2 above)
Comment (MattDavisRV) [reply chat.go]: Probably a bad copy and paste, really can't explain it.  Fixed.

File: examples/messages/schedule_message.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+package main
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/MattDavisRV/slack"
```
Comment (james-lawrence) [inline examples/messages/schedule_message.go]: import needs to be fixed.

File: chat.go [hunk 4]
```diff
@@ -155,6 +171,61 @@ func (api *Client) SendMessage(channel string, options ...MsgOption) (string, st
...
 
+// PostScheduledMessage schedules a messagee to be sent at a later date
+// Use http://davestevens.github.io/slack-message-builder/ to help crafting your message.
+// See https://api.slack.com/messaging/scheduling for more information on scheduling
+func (api *Client) PostScheduledMessage(channelID string, timestamp string, options ...MsgOption) (string, string, error) {
+	respChannel, respTimestamp, err := api.PostScheduledMessageContext(
+		context.Background(),
+		channelID,
+		timestamp,
+		MsgOptionCompose(options...),
```
Comment (james-lawrence) [inline chat.go]: Compose isn't necessary here. its no different than just `options...`

File: chat.go [hunk 5]
```diff
@@ -155,6 +171,61 @@ func (api *Client) SendMessage(channel string, options ...MsgOption) (string, st
...
+}
+
+// PostScheduledMessageContext schedules a message to be sent with custom context
+// See PostScheduledMessage for more information
+func (api *Client) PostScheduledMessageContext(ctx context.Context, channelID string, timestamp string, options ...MsgOption) (string, string, error) {
+	respChannel, respTimestamp, _, err := api.SendMessageContext(
+		ctx,
+		channelID,
+		MsgOptionSchedule(timestamp),
+		MsgOptionCompose(options...),
```
Comment (james-lawrence) [inline chat.go:194]: good usage of compose.

**Review by james-lawrence — CHANGES_REQUESTED** [review]
Overall looks good. minor comments.

main thing missing are tests for scheduled functionality as they use new endpoints.

there is the new slacktest package to look at to help implement the tests.

**Comment by MattDavisRV** [comment]
I've updated this PR based on the feedback.  I will be added tests later today.

**Comment by MattDavisRV** [comment]
> Overall looks good. minor comments.
> 
> main thing missing are tests for scheduled functionality as they use new endpoints.
> 
> there is the new slacktest package to look at to help implement the tests.

@james-lawrence do you want me add handlers to slacktest for the schedule / list / delete for the tests?  If not, I'm not sure how that package will help test the added code.

**Comment by khayyamsaleem** [comment]
@MattDavisRV do you still have an interest in this PR / would you like some help with it?

**Comment by MattDavisRV** [comment]
If you'd like to write tests for it, then that'd be helpful.  I had the outstanding question for @james-lawrence, but I had forgotten about this.  I've been using my fork in production within my environment, but I'd like to see it get merged at some point.

**Pull request closed without merging**